TCPGen implementation

BriansIDP opened the pull request:
This is the implementation of the tree-constrained pointer generator for contextual biasing in ESPNet2 using LibriSpeech train-clean-100 data.

sw005320 commented on egs2/librispeech_100/asr1_biasing/README.md:
- Can you add the model checkpoint? See https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#132-espnet2-recipes
- Can you also add a result of rare word recognition?
  - How to compute it?
  - Can you add this scoring tool in `local/score.sh`?

BriansIDP replied:
Might take a couple of days to finish but I will work on this as soon as possible. Thank you!

sw005320 commented on egs2/librispeech_100/asr1_biasing/conf/decode_asr.yaml:
are you using this?

sw005320 commented on egs2/librispeech_100/asr1_biasing/conf/decode_asr_whisper_noctc_greedy.yaml:
Is this naming OK?
Why whisper?

BriansIDP replied:
Hi Shinji. Thank you for your comments! I think I left a few unnecessary files here and I should clean them up.

sw005320 commented on egs2/librispeech_100/asr1_biasing/db.sh:
This should be replaced with a symlink.

BriansIDP replied:
OK noted.

sw005320 commented on egs2/librispeech_100/asr1_biasing/decode.sh:
Move it under `local`

BriansIDP replied:
Yes, I will remove it and integrate into run.sh. It was just clearer on my side to do debugging stuff with a single script.

sw005320 reviewed (commented):
Thanks a lot!
This is a very great PR!

I have several high-level suggestions. I would like @simpleoier to check more details.
- We need to add [unit test](https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#4-unit-testing) and [integration test](https://github.com/espnet/espnet/blob/master/CONTRIBUTING.md#5-integration-testing)
- This is a very cool function, so we can add more detailed documents. Maybe, we can add a usage in egs2/librispeech_100/asr1_biasing/README.md
- Can you also add a bias list that is used in your experiments? You could add it under the `local` directory.

simpleoier commented on espnet2/bin/asr_inference.py:
```diff
@@ -951,6 +1007,29 @@ def get_parser():
+        default=100,
+        help="maximum biasing list size",
+    )
+    group.add_argument(
```
Do you use biasing list dropout during inference?

BriansIDP replied:
I wanted to keep the option for using dropout during inference, but in my paper, I did not. The task of biasing is that we have a list of words/entities that we want the ASR to perform better on, so we normally assume we know what we want to bias. It is just a matter of how many distractors we get that influence the performance.

simpleoier commented on espnet2/tasks/abs_task.py:
```diff
@@ -846,6 +846,55 @@ class ArgumentDefaultsRawTextHelpFormatter(
             "e.g., 'bias_weight_decay': False will set zero weight decay for bias "
             "params. See also espnet2.optimizers.optim_groups.configure_optimizer.",
         )
+        group.add_argument(
```
The following arguments are task-specific. You don't have to put biasing-related arguments here. Instead, you can put them in the configuration `model_conf`. They will be passed  to the `espnet_model`.

BriansIDP replied:
So actually there are biasing-related arguments that I need to pass to the joint network as well. Is it better to keep these arguments, or is it better to have those in both model_conf and joint_net_conf in the config file?

simpleoier reviewed (commented):
Thanks for your efforts! I went through the scripts once. I put some suggestions and some questions. We can discuss and improve this PR.

BriansIDP commented on espnet2/bin/asr_inference.py:
```diff
@@ -678,6 +722,16 @@ def inference(
+    blist_perutt = None
+    if speech2text.asr_model.biasing and perutt_blist != "":
+        with open(perutt_blist) as fin:
+            blist_perutt = json.load(fin)
```
Yes, the biasing list for the test set is known for each utterance. So for each utterance, we know what are the words in the utterance that we are going to bias. However, in reality, while we can gather biasing words, e.g. restaurant names, we often gather much more than we needed, and one possible simulation is to add distractors. In my setup, we add 1000 distractors during inference in addition to the biasing words from the utterance.

BriansIDP commented on espnet2/train/trainer.py:
```diff
@@ -336,6 +338,13 @@ def run(
             if not distributed_option.distributed or distributed_option.dist_rank == 0:
                 # 3. Report the results
                 logging.info(reporter.log_message())
+                # Hack here for logging
```
Easier for me to debug. Forgot to remove that. Sorry.

BriansIDP commented on egs2/librispeech_100/asr1_biasing/conf/train_rnnt.yaml:
```diff
@@ -0,0 +1,91 @@
+
+# Biasing
+# Change this to the path of your biasing list
+biasinglist: data/Blist/rareword_f15.txt
```
Will add to local. Just choose words appeared fewer than 15 times in the training set.

BriansIDP commented on espnet2/asr/decoder/transducer_decoder.py:
```diff
@@ -59,6 +59,7 @@ def __init__(
         self.dunits = hidden_size
         self.dtype = rnn_type
         self.odim = vocab_size
+        self.embed_size = hidden_size
```
Ah did not notice that. Sorry.

BriansIDP commented on espnet2/asr/transducer/beam_search_transducer.py:
```diff
@@ -367,6 +455,94 @@ def default_beam_search(self, enc_out: torch.Tensor) -> List[Hypothesis]:
 
         return self.sort_nbest(kept_hyps)
 
+    def get_step_biasing_embs(self, char_ids, trees, origTries, node_encs=None):
```
They are separate functions. One is to get biasing indices and one is to perform the forward of the TCPGen component. Perhaps I should rename those.

BriansIDP replied:
Oh yes, they are. I actually tried to import those from one place to another but encountered a looped import error. I will have another go with this.

BriansIDP commented on espnet2/asr_transducer/joint_network.py:
```diff
@@ -25,24 +25,34 @@ def __init__(
 
         self.lin_enc = torch.nn.Linear(encoder_size, joint_space_size)
-        self.lin_dec = torch.nn.Linear(decoder_size, joint_space_size)
+        self.lin_dec = torch.nn.Linear(decoder_size, joint_space_size, bias=False)
```
It's not. I will change it back. It was in my previous version of ESPNet2 which I did not sync. Sorry about it.

BriansIDP replied:
Just realised that I need to train my model again without this. Let me do it now.

BriansIDP commented on espnet2/asr/espnet_model.py:
```diff
@@ -619,16 +672,91 @@ def _calc_transducer_loss(
+        if self.biasing and self.epoch >= self.biasingsche:
+            # Encode prefix tree using GNN
+            if self.GNN != "":
+                node_encs = self.gnn(lextrees[0], self.decoder.embed)
```
Just get the embedding weights (just get the data, no gradient backpropagation) by index-selecting `self.decoder.embed.weight.data`

BriansIDP commented on espnet2/asr/espnet_model.py:
```diff
@@ -670,3 +798,92 @@ def _calc_batch_ctc_loss(
         self.ctc.reduce = do_reduce
         return loss_ctc
+
+    def get_step_biasing_embs(self, char_ids, trees, origTries, node_encs=None):
```
I will add descriptions.

BriansIDP commented on espnet2/tasks/abs_task.py:
```diff
@@ -846,6 +846,55 @@ class ArgumentDefaultsRawTextHelpFormatter(
+            default=False,
+            help="perform biasing",
+        )
+        group.add_argument(
```
biasing means performing biasing with TCPGen, and in my paper, TCPGen can also be combined with deep biasing where the biasing vector, h_ptr, is fed into the joint network. That is, when biasing=true and deepbiasnig=false, this is TCPGen only performing modifications directly to the model output (which yielded slightly worse results than using both TCPGen and deep biasing according to my paper).

simpleoier commented on egs2/librispeech_100/asr1_biasing/README.md:
```diff
@@ -0,0 +1,38 @@
+- espnet version: `espnet 202304`
+- pytorch version: `pytorch 2.0.1+cu117`
+- Git hash: `6f33b9d9a999d4cd7e9bc0dcfc0ba342bdff7c17`
+  - Commit date: `Thu Jun 29 02:16:09 2023 +0100`
```
Will you upload the model to huggingface and put the link here?

BriansIDP replied:
I have the checkpoint ready. Should I do it now or later when the code is checked in? (Also any links or instructions I should follow for submitting?) Thank you!

simpleoier replied:
You can upload it now or later. If the code changes a lot, the ckpt you have now may need to be updated.

sw005320 replied:
@BriansIDP, let's add it.
Please follow the other recipe and put it.

simpleoier commented on egs2/librispeech_100/asr1_biasing/conf/train_rnnt.yaml:
```diff
@@ -0,0 +1,90 @@
+optim: adam
+optim_conf:
+    lr: 0.002
+    # weight_decay: 0.000001
```
Since this hyperparameter is not used, will you delete it?

BriansIDP replied:
Yes I will. Thanks!

simpleoier commented on egs2/librispeech_100/asr1_biasing/asr.sh:
```diff
@@ -0,0 +1,1857 @@
+        rm -f "${_logdir}/*.log"
+        # shellcheck disable=SC2046,SC2086
+        ${_cmd} --gpu "${_ngpu}" JOB=1:"${_nj}" "${_logdir}"/asr_inference.JOB.log \
+            ${python} -m espnet2.bin.${asr_task}_inference${inference_bin_tag} \
```
Probably you can add biasing-related decoding file above. Then you don't need the separate `decode.sh`.

BriansIDP replied:
I will remove it now. The run.sh now can run through the entire pipeline (I am still working on the calculating biasing word error part and will add that in to `asr.sh` as well when it is done.)

simpleoier commented on egs2/librispeech_100/asr1_biasing/decode.sh:
```diff
@@ -0,0 +1,40 @@
+. ./path.sh
+. ./cmd.sh
+export PYTHONPATH="/home/gs534/rds/rds-t2-cs164-KQ4S3rlDzm8/gs534/opensource/espnet:$PYTHONPATH"
```
This path contains your local information. As I suggested above, you can merge this into `asr.sh`.

BriansIDP replied:
Will remove this decode.sh file in a later push.

simpleoier commented on egs2/librispeech_100/asr1_biasing/local/error_analysis/get_error_word_count.py:
This file has many unused lines.

BriansIDP replied:
Will work on this and make it clean, put it under local/ and write something in asr.sh to execute it.

simpleoier commented on egs2/librispeech_100/asr1_biasing/run.sh:
```diff
@@ -0,0 +1,40 @@
+set -u
+set -o pipefail
+
+export PATH=~/rds/rds-t2-cs164-KQ4S3rlDzm8/gs534/MultiModal/espnet/tools/sctk/bin/:$PATH
```
This PATH contains some local information. I'm not sure. Probably sctk does not require particular `export`.

BriansIDP replied:
My local env is a bit complicated so I will remove it once the scoring with biasing words is done. Thanks!

simpleoier reviewed (commented):
I added some more comments.

BriansIDP commented:
Hi @sw005320 @simpleoier @emonosuke , I think I have addressed most of the comments except for the unit-test. I will upload the model to huggingface now and will go through your comments again this week. Thank you!

BriansIDP commented:
> Hi @sw005320 @simpleoier @emonosuke , I think I have addressed most of the comments except for the unit-test. I will upload the model to huggingface now and will go through your comments again this week. Thank you!

I have uploaded the model now named `tcpgen_biasing` under `espnet`. As this is my first time pushing a model to huggingface, I am not sure how I could verify "using the inference demo". Any help would be appreciated! Thank you!

simpleoier commented:
@BriansIDP You need to follow stage 14 & 16 in `asr.sh`.

BriansIDP commented:
> 

Thanks for the pointer. I have done following the instructions in stage 14 & 16 and the model now is at: 
[tcpgen_biasing](https://huggingface.co/espnet/tcpgen_biasing).

simpleoier commented:
@BriansIDP Another suggestion: it may be better to update the huggingface repo name. If you notice other pretrained checkpoints, you can at least find the name follows the pattern: contributer_dataset_expname.

BriansIDP commented:
> @BriansIDP Another suggestion: it may be better to update the huggingface repo name. If you notice other pretrained checkpoints, you can at least find the name follows the pattern: contributer_dataset_expname.

Thanks. I've done that.

sw005320 commented on egs2/librispeech_100/asr1_biasing/asr.sh:
```diff
@@ -0,0 +1,1849 @@
+#!/usr/bin/env bash
```
I agree.

@BriansIDP, with this change, your work becomes a template!
I think it's worth doing this effort.

sw005320 commented on egs2/librispeech_100/asr1_biasing/conf/train_rnnt.yaml:
```diff
@@ -0,0 +1,89 @@
+# Trained with A40 (48 GB) x 4 GPUs.
```
Can you highlight (add a comment) lines specific to tcpgen?

sw005320 commented on egs2/librispeech_100/asr1_biasing/conf/tuning/train_rnnt_std_tcpgen.yaml:
```diff
@@ -0,0 +1,90 @@
+# Trained with A40 (48 GB) x 4 GPUs.
```
ditto
Can you highlight (add a comment) lines specific to tcpgen?

sw005320 commented on espnet2/asr/espnet_model.py:
```diff
@@ -1,15 +1,18 @@
 import logging
```
I agree.

espnet_model.py is getting larger, and I want to avoid it to be further larger.

sw005320 commented on espnet2/asr/layers/gnn.py:
```diff
@@ -0,0 +1,137 @@
+from __future__ import division
```
I agree.

sw005320 commented on espnet2/asr/transducer/beam_search_transducer.py:
```diff
@@ -173,7 +194,11 @@ def __call__(
         self.decoder.set_device(enc_out.device)
 
-        nbest_hyps = self.search_algorithm(enc_out)
+        # Only supporting biasing in default_beam_search for now
```
I prefer a new function since this is new (novel)!

sw005320 reviewed (commented):
@BriansIDP, can you deal with @simpleoier's comments?
They are very reasonable to me.

sw005320 commented:
Another comment is that the file list is too large to be maintained as a text file in Git.
We may change it to the compressed (zip) file, but the current way is fine if this is complicated.

sw005320 commented:
We could upload a model later.
A valid source is more important.

BriansIDP commented:
Hi Shinji @sw005320 . Sorry for the late reply! I have modified the model file and the beam search so that they are now separate files. I have also trained a model which yielded similar results. I will upload the model soon but the code review could continue on the newly added files. Thank you!

sw005320 commented:
Thanks!
There are several conflicts. So, can you fix them?
Then, we'll have another round of reviews.

I really want to merge this PR. So, let's work together to refine it.

BriansIDP commented:
> Thanks! There are several conflicts. So, can you fix them? Then, we'll have another round of reviews.
> 
> I really want to merge this PR. So, let's work together to refine it.

Thank you. I have now resolved the conflicts.

BriansIDP commented on espnet2/bin/asr_inference.py:
```diff
@@ -210,6 +238,9 @@ def __init__(
                 multi_blank_indices=multi_blank_indices,
                 token_list=token_list,
+                biasing=asr_model.biasing,
+                deepbiasing=asr_model.deepbiasing,
```
Thanks! I have pushed the changes and I guess this is the only place where this caused a problem. Please let me know if I missed anything. Thank you!

BriansIDP commented:
> Thanks! There are several conflicts. So, can you fix them? Then, we'll have another round of reviews.
> 
> I really want to merge this PR. So, let's work together to refine it.

Hi Shinji. I wonder if there will be another round of review and what the further steps for this could be. It would be greatly appreciated if this could be integrated into ESPNet. Thank you!

simpleoier commented on egs2/TEMPLATE/asr1/asr.sh:
```diff
@@ -906,6 +919,7 @@ if [ ${stage} -le 5 ] && [ ${stop_stage} -ge 5 ] && ! [[ " ${skip_stages} " =~ [
             --model_type="${bpemode}" \
             --model_prefix="${bpeprefix}" \
             --character_coverage=${bpe_char_cover} \
+            --treat_whitespace_as_suffix=${usesuffixbpe} \
```
```suggestion
```

simpleoier replied:
Instead, you may add the following above
```
if [ -n "${suffixbpe}" ]; then
    _opts_spm+="--treat_whitespace_as_suffix=${usesuffixbpe} "
fi
```

sw005320 replied:
This one (removing the existing one) would break compatibility. Please follow @simpleoier's suggestion.

simpleoier commented on espnet2/asr_transducer/joint_network.py:
```diff
@@ -41,11 +44,18 @@ def __init__(
         self,
         enc_out: torch.Tensor,
         dec_out: torch.Tensor,
+        hptr: torch.Tensor = None,
```
a bit concern is that `hptr` is before `no_projection`. Will this order change affect other transducer models?

BriansIDP replied:
Thanks for the comments! I have now addressed the problem in asr.sh, and for this, I moved `hptr` after `no_projection` and changed various places where this forward is invoked.

sw005320 commented on espnet2/asr/TCPGen_biasing_model.py:
Please follow the naming convention.
espnet2/asr/TCPGen_biasing_model.py --> espnet2/asr/tcpgen_biasing_espnet_model.py

sw005320 commented on espnet2/asr/transducer/beam_search_transducer.py:
```diff
@@ -379,6 +406,263 @@ def default_beam_search(self, enc_out: torch.Tensor) -> List[Hypothesis]:
 
         return self.sort_nbest(kept_hyps)
 
+    def TCPGen_biasing_beam_search(
```
```suggestion
    def tcpgen_biasing_beam_search(
```

sw005320 commented on espnet2/bin/asr_inference.py:
```diff
@@ -373,7 +404,9 @@ def __init__(
-        self, speech: Union[torch.Tensor, np.ndarray]
+        self,
+        speech: Union[torch.Tensor, np.ndarray],
+        blist: list = None,
```
@BriansIDP, please reflect on it.

sw005320 commented on espnet2/bin/asr_inference.py:
```diff
@@ -466,10 +503,10 @@ def _decode_interctc(
         return res
 
-    def _decode_single_sample(self, enc: torch.Tensor):
+    def _decode_single_sample(self, enc: torch.Tensor, lextree: list = None):
```
ditto

sw005320 commented:
Please fix the CI error https://github.com/espnet/espnet/actions/runs/6710579965/job/18236089339?pr=5261#step:8:259

sw005320 commented:
This PR requires a unit test. Please check https://github.com/espnet/espnet/tree/master/test/espnet2. You can find a lot of examples.

BriansIDP commented:
Hi Shinji @sw005320. I have written some unittest for my newly added code, but they failed because "warp-rnnt" is not installed. I noticed that the current version of ESPNet does not support this library. I wonder what the reason is, and what I should do. Thank you!

sw005320 commented:
We are using https://github.com/espnet/espnet/blob/master/tools/installers/install_warp-transducer.sh#L29, which will be installed in our normal CI configuration in https://github.com/espnet/espnet/blob/master/ci/install.sh#L24
Can. you double-check it?

BriansIDP commented:
> We are using https://github.com/espnet/espnet/blob/master/tools/installers/install_warp-transducer.sh#L29, which will be installed in our normal CI configuration in https://github.com/espnet/espnet/blob/master/ci/install.sh#L24 Can. you double-check it?

Hi Shinji. Yes, I understand that `warp-transducer` is used. However, `warp-transducer` does not support log probabilities as input whereas the output of TCPGen performs distribution-level interpolation and hence requires log probabilities to be sent to the transducer loss. Therefore I used `warp-rnnt` instead. I remember that in the older versions of ESPNet, `warp-rnnt` was supported. I wonder if we could make use of that installer. Thank you!

sw005320 commented:
I see.
We removed warp-rnnt in https://github.com/espnet/espnet/pull/3217.
@b-flo, do you think we can update the interface to support it in warp-transducer?

b-flo commented:
Hi,

> @b-flo, do you think we can update the interface to support it in warp-transducer?

Sure, let me take a look later this week. For the GPU version the log softmax is computed inside the implementation, I could add an option to make it optional I guess.

That being said, it seems the implementation is for both `ASRTask` with transducer mode and `ASRTransducerTask`, is it intended? For the later, k2 transducer loss can be used instead (with the proper `k2_pruned_loss_args`)

BriansIDP commented:
> Hi,
> 
> > @b-flo, do you think we can update the interface to support it in warp-transducer?
> 
> Sure, let me take a look later this week. For the GPU version the log softmax is computed inside the implementation, I could add an option to make it optional I guess.
> 
> That being said, it seems the implementation is for both `ASRTask` with transducer mode and `ASRTransducerTask`, is it intended? For the later, k2 transducer loss can be used instead (with the proper `k2_pruned_loss_args`)

Hi Florian. Thank you for the reply! Actually the implementation currently is only for ASRTask with transducer model.

b-flo commented:
> Actually the implementation currently is only for ASRTask with transducer model.

Got it sorry, the decoder and joint network are shared across version. I thought I made the task/models fully separated in the past.
Hum, we should maybe define another `JointNetwork` for ASRTask to avoid dead codes for each version. Let me check alongside the other problem.

BriansIDP commented:
> > Actually the implementation currently is only for ASRTask with transducer model.
> 
> Got it sorry, the decoder and joint network are shared across version. I thought I made the task/models fully separated in the past. Hum, we should maybe define another `JointNetwork` for ASRTask to avoid dead codes for each version. Let me check alongside the other problem.

Hi @b-flo, I wonder if there is any update on this problem. Thank you!
